# AUDIT RULE: Financial (Financeiro)

## üéØ Objetivo
Garantir que m√≥dulos com opera√ß√µes financeiras sigam os princ√≠pios de **idempot√™ncia**, **auditabilidade** e **integridade de dados**.

---

## ‚úÖ Checklist de Auditoria

### 1. **Idempot√™ncia**
- [ ] Transa√ß√µes usam identificadores √∫nicos (`chaveIdempotencia`)
- [ ] Opera√ß√µes financeiras previnem cobran√ßa duplicada
- [ ] Usa `findOneAndUpdate` com `upsert` quando aplic√°vel
- [ ] Valida se transa√ß√£o j√° existe antes de criar

**Exemplo correto:**
```javascript
const chaveIdempotencia = `${tipo}-${timeId}-${temporada}-${identificador}`;
const existente = await ExtratoFinanceiro.findOne({ chaveIdempotencia });
if (existente) {
    return { ja_existe: true };
}
```

---

### 2. **Registro no Extrato**
- [ ] Toda transa√ß√£o financeira gera entrada em `extratofinanceiro`
- [ ] Campos obrigat√≥rios preenchidos: `tipo`, `valor`, `descricao`, `temporada`, `timeId`
- [ ] Campo `chaveIdempotencia` presente e √∫nico
- [ ] Data registrada (`data` ou `timestamp`)

**Campos cr√≠ticos:**
```javascript
{
    tipo: 'debito' | 'credito',
    valor: Number,
    descricao: String,
    temporada: Number,
    timeId: String | Number,
    chaveIdempotencia: String,
    data: Date
}
```

---

### 3. **Valida√ß√£o de Sess√£o**
- [ ] Valida `req.session.usuario` antes de opera√ß√µes sens√≠veis
- [ ] Verifica autoriza√ß√£o admin quando necess√°rio (`isAdminAuthorizado`)
- [ ] Previne manipula√ß√£o de dados de outros usu√°rios

**Exemplo correto:**
```javascript
if (!req.session.usuario) {
    return res.status(401).json({ erro: 'N√£o autorizado' });
}

// Para admin
if (!isAdminAutorizado(req.session.usuario.email)) {
    return res.status(403).json({ erro: 'Acesso negado' });
}
```

---

### 4. **Opera√ß√µes At√¥micas**
- [ ] Usa MongoDB atomic operations (`$inc`, `$set`, `$push`)
- [ ] Evita race conditions (read-modify-write)
- [ ] Considera usar transactions para m√∫ltiplas opera√ß√µes

**Exemplo correto:**
```javascript
// ‚ùå Evitar (race condition)
const saldo = await calcularSaldo(timeId);
await salvarSaldo(timeId, saldo + valor);

// ‚úÖ Correto (at√¥mico)
await ExtratoCache.findOneAndUpdate(
    { timeId, temporada },
    { $inc: { saldo: valor } },
    { upsert: true }
);
```

---

### 5. **Auditoria "Follow the Money"**
- [ ] Transa√ß√µes t√™m rastreabilidade (origem/destino)
- [ ] Logs adequados para opera√ß√µes financeiras cr√≠ticas
- [ ] Campo `metadata` ou `detalhes` com contexto adicional

**Exemplo:**
```javascript
{
    tipo: 'debito',
    valor: 10,
    descricao: 'Aposta Artilheiro - Cristiano Ronaldo',
    metadata: {
        modulo: 'artilheiro',
        jogador_id: 12345,
        rodada: 15
    }
}
```

---

### 6. **Separa√ß√£o por Temporada**
- [ ] Extratos separados por temporada (campo `temporada`)
- [ ] N√£o mistura saldos de temporadas diferentes
- [ ] Sistema de "legado" implementado corretamente

---

### 7. **Tratamento de Erros**
- [ ] `try/catch` em todas fun√ß√µes async
- [ ] Rollback em caso de falha (quando aplic√°vel)
- [ ] Mensagens de erro claras para o usu√°rio

**Exemplo:**
```javascript
try {
    await ExtratoFinanceiro.create({ ... });
} catch (erro) {
    console.error('Erro ao criar extrato:', erro);
    return res.status(500).json({
        erro: 'Falha ao processar transa√ß√£o',
        detalhes: process.env.NODE_ENV === 'development' ? erro.message : undefined
    });
}
```

---

### 8. **Tipos de ID Consistentes**
- [ ] Respeita tipo de ID por collection (ver CLAUDE.md)
- [ ] `extratofinanceirocaches`: `time_id` (Number)
- [ ] `fluxofinanceirocampos`: `timeId` (String)
- [ ] `acertofinanceiros`: `timeId` (String)

---

### 9. **Flag `pagouInscricao`**
- [ ] Inscri√ß√µes de temporada usam flag corretamente
- [ ] `true` = taxa registrada (n√£o vira d√©bito)
- [ ] `false` = taxa vira d√©bito no extrato
- [ ] Integra√ß√£o com `inscricoestemporada` collection

---

### 10. **Cache Financeiro**
- [ ] Usa `extratofinanceirocaches` para performance
- [ ] Cache invalidado ap√≥s transa√ß√µes
- [ ] Rec√°lculo via endpoint `/api/extrato-financeiro/recalcular-cache`

---

## üö® Red Flags Cr√≠ticos

| Problema | Severidade | A√ß√£o |
|----------|-----------|------|
| Sem `chaveIdempotencia` | üî¥ CR√çTICO | Implementar imediatamente |
| Opera√ß√µes n√£o at√¥micas | üî¥ CR√çTICO | Refatorar para atomic ops |
| Sem valida√ß√£o de sess√£o | üî¥ CR√çTICO | Adicionar autoriza√ß√£o |
| Extrato sem temporada | üü† ALTO | Adicionar campo |
| Sem try/catch | üü† ALTO | Adicionar error handling |
| Cache desatualizado | üü° M√âDIO | Invalidar ap√≥s transa√ß√µes |

---

## üìä Exemplos de Uso

### ‚úÖ Exemplo Completo (Aposta Artilheiro)
```javascript
async function criarAposta(req, res) {
    try {
        // 1. Validar sess√£o
        if (!req.session.usuario) {
            return res.status(401).json({ erro: 'N√£o autorizado' });
        }

        const { timeId, jogadorId, temporada, valor } = req.body;

        // 2. Chave de idempot√™ncia
        const chaveIdempotencia = `aposta-artilheiro-${timeId}-${temporada}-${jogadorId}`;

        // 3. Verificar duplicata
        const existente = await ExtratoFinanceiro.findOne({ chaveIdempotencia });
        if (existente) {
            return res.status(409).json({ erro: 'Aposta j√° realizada' });
        }

        // 4. Registrar no extrato (opera√ß√£o at√¥mica)
        await ExtratoFinanceiro.create({
            tipo: 'debito',
            valor: valor,
            descricao: `Aposta Artilheiro - Jogador ${jogadorId}`,
            temporada: temporada,
            timeId: timeId,
            chaveIdempotencia: chaveIdempotencia,
            data: new Date(),
            metadata: {
                modulo: 'artilheiro',
                jogador_id: jogadorId
            }
        });

        // 5. Invalidar cache
        await invalidarCacheFinanceiro(timeId, temporada);

        res.json({ sucesso: true });

    } catch (erro) {
        console.error('Erro ao criar aposta:', erro);
        res.status(500).json({ erro: 'Falha ao processar aposta' });
    }
}
```

---

## üîó Refer√™ncias
- `CLAUDE.md` ‚Üí Se√ß√£o "Coding Standards" e "Estrutura de Dados"
- `docs/SISTEMA-RENOVACAO-TEMPORADA.md` ‚Üí Flags e terminologia
- `controllers/extrato-financeiro-controller.js` ‚Üí Implementa√ß√£o de refer√™ncia

---

**√öltima atualiza√ß√£o:** 04/02/2026
**Vers√£o:** 1.0.0
