# AUDIT RULE: Performance (OtimizaÃ§Ã£o)

## ğŸ¯ Objetivo
Garantir que mÃ³dulos sejam **eficientes**, **escalÃ¡veis** e nÃ£o causem **lentidÃ£o** ou **timeouts** em produÃ§Ã£o.

---

## âœ… Checklist de Auditoria

### 1. **Queries MongoDB Otimizadas**

#### Ãndices
- [ ] Collections tÃªm Ã­ndices nos campos mais consultados
- [ ] Queries complexas usam Ã­ndices compostos
- [ ] `explain()` executado para validar uso de Ã­ndice

**Verificar Ã­ndices:**
```javascript
db.extratofinanceiro.getIndexes()
```

**Criar Ã­ndices:**
```javascript
// Ãndice simples
db.extratofinanceiro.createIndex({ temporada: 1 })

// Ãndice composto
db.extratofinanceiro.createIndex({ timeId: 1, temporada: -1 })

// Ãndice Ãºnico (idempotÃªncia)
db.extratofinanceiro.createIndex({ chaveIdempotencia: 1 }, { unique: true })
```

#### ProjeÃ§Ã£o de Campos
- [ ] Usa `.select()` para buscar apenas campos necessÃ¡rios
- [ ] Evita carregar documentos completos desnecessariamente

**âŒ INEFICIENTE:**
```javascript
const participantes = await Time.find({ ligaId }); // carrega tudo
```

**âœ… OTIMIZADO:**
```javascript
const participantes = await Time.find({ ligaId })
    .select('nome_time escudo_id pontos') // apenas campos necessÃ¡rios
    .lean(); // retorna POJO (nÃ£o Mongoose document)
```

#### PaginaÃ§Ã£o
- [ ] Listas longas usam paginaÃ§Ã£o (`.skip()` + `.limit()`)
- [ ] Limite mÃ¡ximo por pÃ¡gina (ex: 50 itens)
- [ ] Cursor-based pagination para grandes volumes

**Exemplo:**
```javascript
const page = parseInt(req.query.page) || 1;
const limit = 50;
const skip = (page - 1) * limit;

const resultados = await Documento.find(query)
    .skip(skip)
    .limit(limit);
```

---

### 2. **Cache EstratÃ©gico**

#### Quando Cachear
- [ ] Dados que mudam pouco (config de liga, regras)
- [ ] CÃ¡lculos pesados (saldo financeiro, rankings)
- [ ] Respostas de APIs externas (Cartola API)

**Exemplo (Redis ou in-memory):**
```javascript
const cacheKey = `saldo-${timeId}-${temporada}`;
let saldo = cache.get(cacheKey);

if (!saldo) {
    saldo = await calcularSaldoCompleto(timeId, temporada);
    cache.set(cacheKey, saldo, 3600); // 1 hora
}
```

#### InvalidaÃ§Ã£o de Cache
- [ ] Cache invalidado apÃ³s updates relevantes
- [ ] TTL (Time To Live) configurado adequadamente
- [ ] Evita cache stale crÃ­tico

**Exemplo:**
```javascript
async function criarTransacao(dados) {
    await ExtratoFinanceiro.create(dados);

    // Invalidar cache
    const cacheKey = `saldo-${dados.timeId}-${dados.temporada}`;
    cache.del(cacheKey);
}
```

#### Cache Collections
- [ ] `extratofinanceirocaches` usado corretamente
- [ ] Endpoint `/recalcular-cache` disponÃ­vel
- [ ] Cache atualizado apÃ³s mudanÃ§as financeiras

---

### 3. **Queries N+1 (Evitar)**

#### Problema
Executar query dentro de loop â†’ N queries adicionais.

**âŒ N+1 PROBLEM:**
```javascript
const apostas = await Aposta.find({ temporada });
for (const aposta of apostas) {
    // QUERY DENTRO DO LOOP!
    const time = await Time.findOne({ id: aposta.timeId });
    aposta.nomeTime = time.nome_time;
}
```

**âœ… SOLUÃ‡ÃƒO (bulk fetch):**
```javascript
const apostas = await Aposta.find({ temporada });
const timeIds = apostas.map(a => a.timeId);

// UMA query para todos times
const times = await Time.find({ id: { $in: timeIds } });
const timesMap = new Map(times.map(t => [t.id, t]));

apostas.forEach(aposta => {
    aposta.nomeTime = timesMap.get(aposta.timeId)?.nome_time;
});
```

---

### 4. **Frontend: Lazy Loading**

#### Carregamento Sob Demanda
- [ ] Imagens usam `loading="lazy"`
- [ ] Listas longas renderizam virtualizadas
- [ ] JS/CSS carregados assincronamente quando possÃ­vel

**HTML:**
```html
<img src="/escudos/262.png" loading="lazy" alt="Flamengo">
```

**JavaScript (Intersection Observer):**
```javascript
const observer = new IntersectionObserver((entries) => {
    entries.forEach(entry => {
        if (entry.isIntersecting) {
            entry.target.src = entry.target.dataset.src;
            observer.unobserve(entry.target);
        }
    });
});

document.querySelectorAll('img[data-src]').forEach(img => {
    observer.observe(img);
});
```

---

### 5. **Reduzir Payload (Response Size)**

#### Apenas Dados NecessÃ¡rios
- [ ] API retorna apenas campos usados pelo frontend
- [ ] Remove metadados desnecessÃ¡rios (`__v`, `_id` interno)
- [ ] CompressÃ£o gzip ativada no servidor

**Exemplo:**
```javascript
const participantes = await Time.find({ ligaId })
    .select('nome_time pontos escudo_id') // sÃ³ o necessÃ¡rio
    .lean(); // remove metadados Mongoose

res.json(participantes); // payload menor
```

---

### 6. **Debounce/Throttle em Inputs**

#### Busca em Tempo Real
- [ ] Inputs de busca usam debounce (espera digitaÃ§Ã£o terminar)
- [ ] Scroll infinito usa throttle (limita chamadas)

**Exemplo (debounce):**
```javascript
let timeoutId;
const inputBusca = document.getElementById('busca');

inputBusca.addEventListener('input', (e) => {
    clearTimeout(timeoutId);
    timeoutId = setTimeout(() => {
        buscarJogadores(e.target.value);
    }, 300); // aguarda 300ms apÃ³s Ãºltima tecla
});
```

---

### 7. **AgregaÃ§Ãµes MongoDB**

#### Pipeline Aggregation
- [ ] Usa `.aggregate()` para cÃ¡lculos complexos no banco
- [ ] Evita trazer dados para JS e processar lÃ¡

**Exemplo (saldo por temporada):**
```javascript
const saldos = await ExtratoFinanceiro.aggregate([
    { $match: { timeId: '123', temporada: 2026 } },
    {
        $group: {
            _id: '$tipo',
            total: { $sum: '$valor' }
        }
    }
]);

const creditos = saldos.find(s => s._id === 'credito')?.total || 0;
const debitos = saldos.find(s => s._id === 'debito')?.total || 0;
const saldo = creditos - debitos;
```

---

### 8. **Async/Await Paralelo**

#### RequisiÃ§Ãµes Independentes
- [ ] Usa `Promise.all()` para requests paralelos
- [ ] Evita `await` sequencial desnecessÃ¡rio

**âŒ LENTO (sequencial):**
```javascript
const liga = await Liga.findOne({ _id: ligaId });
const config = await ModuleConfig.findOne({ modulo: 'top10' });
const participantes = await Time.find({ ligaId });
// Total: tempo1 + tempo2 + tempo3
```

**âœ… RÃPIDO (paralelo):**
```javascript
const [liga, config, participantes] = await Promise.all([
    Liga.findOne({ _id: ligaId }),
    ModuleConfig.findOne({ modulo: 'top10' }),
    Time.find({ ligaId })
]);
// Total: max(tempo1, tempo2, tempo3)
```

---

### 9. **Timeouts e Limites**

#### ProteÃ§Ã£o contra Travamentos
- [ ] Queries tÃªm timeout configurado (ex: 10s)
- [ ] Limita tamanho de arrays retornados (ex: mÃ¡x 1000 itens)
- [ ] API externa tem retry com backoff

**Mongoose timeout:**
```javascript
const resultado = await Documento.find(query)
    .maxTimeMS(10000); // 10 segundos
```

**Retry com backoff:**
```javascript
async function fetchComRetry(url, tentativas = 3) {
    for (let i = 0; i < tentativas; i++) {
        try {
            return await fetch(url);
        } catch (erro) {
            if (i === tentativas - 1) throw erro;
            await new Promise(resolve => setTimeout(resolve, 2 ** i * 1000));
        }
    }
}
```

---

### 10. **Monitoramento de Performance**

#### MÃ©tricas
- [ ] Tempo de resposta dos endpoints logado
- [ ] Queries lentas identificadas (MongoDB slow query log)
- [ ] Memory leaks detectados (heap snapshots)

**Exemplo (tempo de resposta):**
```javascript
router.get('/api/top10', async (req, res) => {
    const inicio = Date.now();

    try {
        const resultado = await buscarTop10();
        res.json(resultado);
    } finally {
        const duracao = Date.now() - inicio;
        if (duracao > 1000) {
            console.warn(`âš ï¸ Endpoint lento: /api/top10 levou ${duracao}ms`);
        }
    }
});
```

---

## ğŸš¨ Red Flags CrÃ­ticos

| Problema | Severidade | Impacto | AÃ§Ã£o |
|----------|-----------|---------|------|
| Query sem Ã­ndice (full scan) | ğŸ”´ CRÃTICO | LentidÃ£o severa | Criar Ã­ndice |
| N+1 queries em loop | ğŸ”´ CRÃTICO | Timeout em prod | Bulk fetch |
| Sem paginaÃ§Ã£o (retorna 10k+ docs) | ğŸ”´ CRÃTICO | Memory overflow | Adicionar paginaÃ§Ã£o |
| Await sequencial de requests independentes | ğŸŸ  ALTO | LatÃªncia 3x maior | Usar Promise.all |
| Cache nunca invalidado | ğŸŸ  ALTO | Dados stale | Implementar invalidaÃ§Ã£o |
| Sem debounce em busca | ğŸŸ¡ MÃ‰DIO | Requests excessivos | Adicionar debounce |
| Response com campos nÃ£o usados | ğŸŸ¡ MÃ‰DIO | Payload inchado | Usar .select() |

---

## ğŸ“Š Benchmark de ReferÃªncia

### Tempos AceitÃ¡veis (95th percentile)

| OperaÃ§Ã£o | Target | Limite |
|----------|--------|--------|
| Query simples (indexed) | < 50ms | 200ms |
| CÃ¡lculo financeiro (cached) | < 100ms | 500ms |
| Endpoint API | < 500ms | 2s |
| Rendering de pÃ¡gina | < 1s | 3s |
| Busca com agregaÃ§Ã£o | < 1s | 5s |

---

## ğŸ”— Ferramentas de AnÃ¡lise

### MongoDB
```javascript
// Analisar query
db.extratofinanceiro.find({ timeId: '123' }).explain('executionStats')

// Ãndices sugeridos
db.extratofinanceiro.aggregate([{ $indexStats: {} }])

// Queries lentas
db.setProfilingLevel(1, { slowms: 100 })
db.system.profile.find().sort({ ts: -1 }).limit(10)
```

### Node.js
```bash
# Memory leaks
node --inspect server.js
# Abrir chrome://inspect

# CPU profiling
node --prof server.js
node --prof-process isolate-*.log > profile.txt
```

---

## ğŸ“š Exemplo Completo (Endpoint Otimizado)

```javascript
router.get('/api/top10/:temporada/:rodada', async (req, res) => {
    const inicio = Date.now();

    try {
        const { temporada, rodada } = req.params;
        const ligaId = req.session.usuario.ligaId;

        // 1. Cache (validar antes de queries pesadas)
        const cacheKey = `top10-${ligaId}-${temporada}-${rodada}`;
        const cached = cache.get(cacheKey);
        if (cached) return res.json(cached);

        // 2. Queries paralelas (Promise.all)
        const [participantes, pontuacoes] = await Promise.all([
            Time.find({ ligaId, temporada: parseInt(temporada), ativo: true })
                .select('id nome_time escudo_id') // apenas campos necessÃ¡rios
                .lean(), // POJO (mais rÃ¡pido)

            Pontuacao.find({
                temporada: parseInt(temporada),
                rodada: parseInt(rodada)
            })
            .select('time_id pontos')
            .sort({ pontos: -1 })
            .limit(10) // paginaÃ§Ã£o
            .lean()
        ]);

        // 3. Join em memÃ³ria (evitar N+1)
        const timesMap = new Map(participantes.map(t => [t.id, t]));

        const top10 = pontuacoes.map((p, idx) => ({
            posicao: idx + 1,
            time: timesMap.get(p.time_id)?.nome_time,
            pontos: p.pontos,
            escudo: timesMap.get(p.time_id)?.escudo_id
        }));

        // 4. Cachear resultado (TTL 1 hora)
        cache.set(cacheKey, top10, 3600);

        res.json(top10);

    } catch (erro) {
        console.error('Erro ao buscar Top 10:', erro);
        res.status(500).json({ erro: 'Falha ao carregar ranking' });
    } finally {
        const duracao = Date.now() - inicio;
        if (duracao > 1000) {
            console.warn(`âš ï¸ /api/top10 demorou ${duracao}ms`);
        }
    }
});
```

---

**Ãšltima atualizaÃ§Ã£o:** 04/02/2026
**VersÃ£o:** 1.0.0
