# AUDIT RULE: UI/UX (Interface e Experi√™ncia)

## üéØ Objetivo
Garantir conformidade com os padr√µes visuais do projeto: **Dark Mode First**, **tipografia consistente** e **identidade visual por m√≥dulo**.

---

## ‚úÖ Checklist de Auditoria

### 1. **Dark Mode (Obrigat√≥rio)**
- [ ] Background principal: `bg-gray-900` ou `bg-slate-900`
- [ ] Cards/Containers: `bg-gray-800`
- [ ] Texto prim√°rio: `text-white` ou `text-gray-100`
- [ ] Texto secund√°rio: `text-gray-400`
- [ ] Inputs: `bg-gray-700 text-white border-gray-600`
- [ ] NUNCA usa cores claras de fundo

**Exemplo correto:**
```html
<div class="bg-gray-900 min-h-screen">
    <div class="bg-gray-800 rounded-lg shadow-lg p-6">
        <h2 class="text-white text-2xl">T√≠tulo</h2>
        <p class="text-gray-400">Descri√ß√£o</p>
    </div>
</div>
```

---

### 2. **Tipografia (3 Fontes Espec√≠ficas)**

| Uso | Fonte | Classe CSS |
|-----|-------|-----------|
| **T√≠tulos, Badges, Stats** | Russo One | `font-russo` |
| **Corpo de texto** | Inter | `font-inter` |
| **Valores num√©ricos** | JetBrains Mono | `font-jetbrains` |

- [ ] T√≠tulos principais usam Russo One
- [ ] Texto corrido usa Inter
- [ ] N√∫meros/estat√≠sticas usam JetBrains Mono
- [ ] Font-face declarado corretamente

**Exemplo correto:**
```html
<h1 class="font-russo text-3xl">Artilheiro Campe√£o</h1>
<p class="font-inter">Sistema de apostas</p>
<span class="font-jetbrains text-2xl">120 pts</span>
```

---

### 3. **Cores dos M√≥dulos (Vari√°veis CSS)**

**üö® REGRA CR√çTICA:** NUNCA usar cores hardcoded (`#22c55e`). SEMPRE usar vari√°veis CSS.

| M√≥dulo | Vari√°vel CSS | Cor |
|--------|--------------|-----|
| Artilheiro | `var(--module-artilheiro-primary)` | Verde `#22c55e` |
| Capit√£o de Luxo | `var(--module-capitao-primary)` | Roxo `#8b5cf6` |
| Luva de Ouro | `var(--module-luva-primary)` | Dourado `#ffd700` |

- [ ] M√≥dulo usa vari√°veis CSS (n√£o hex direto)
- [ ] Background gradiente usa `var(--gradient-[modulo])`
- [ ] Borders usam `var(--module-[modulo]-border)`
- [ ] Fundos sutis usam `var(--module-[modulo]-muted)`

**‚ùå ERRADO:**
```css
.artilheiro-header {
    background: #22c55e;
}
```

**‚úÖ CORRETO:**
```css
.artilheiro-header {
    background: var(--gradient-artilheiro);
    border: 1px solid var(--module-artilheiro-border);
}
```

**Localiza√ß√£o:** `/css/_admin-tokens.css`

---

### 4. **Componentes Padr√£o**

#### Cards
- [ ] `rounded-lg shadow-lg` aplicados
- [ ] Padding adequado (`p-4`, `p-6`)
- [ ] Hover states definidos

```html
<div class="bg-gray-800 rounded-lg shadow-lg p-6 hover:bg-gray-750 transition">
    <!-- Conte√∫do -->
</div>
```

#### Bot√µes
- [ ] Estados hover/active definidos
- [ ] Feedback visual expl√≠cito
- [ ] Cores acess√≠veis (contraste m√≠nimo)

```html
<button class="bg-blue-600 hover:bg-blue-700 active:bg-blue-800
               text-white px-4 py-2 rounded transition">
    A√ß√£o
</button>
```

#### Inputs
- [ ] Fundo escuro (`bg-gray-700`)
- [ ] Texto branco (`text-white`)
- [ ] Border sutil (`border-gray-600`)
- [ ] Focus state definido

```html
<input type="text"
       class="bg-gray-700 text-white border border-gray-600
              rounded px-3 py-2 focus:ring-2 focus:ring-blue-500">
```

---

### 5. **Responsividade Mobile**
- [ ] Breakpoints Tailwind usados (`sm:`, `md:`, `lg:`)
- [ ] Layout adapt√°vel para telas pequenas
- [ ] Texto leg√≠vel em mobile (tamanhos m√≠nimos)
- [ ] Bot√µes com √°rea de toque adequada (min 44x44px)

```html
<div class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
    <!-- Responsive grid -->
</div>
```

---

### 6. **√çcones e Escudos**

#### Escudos de Times
- [ ] Path correto: `/escudos/{clube_id}.png`
- [ ] Fallback implementado: `onerror="this.src='/escudos/default.png'"`
- [ ] Alt text descritivo

```html
<img src="/escudos/262.png"
     alt="Flamengo"
     onerror="this.src='/escudos/default.png'">
```

#### √çcones
- [ ] SVG inline ou biblioteca consistente (Font Awesome, Heroicons)
- [ ] Cores alinhadas com tema dark
- [ ] Tamanho proporcional ao contexto

---

### 7. **Estados Visuais**

#### Loading States
- [ ] Spinners/skeletons para carregamento
- [ ] Feedback visual durante opera√ß√µes async

#### Empty States
- [ ] Mensagem clara quando n√£o h√° dados
- [ ] Sugest√£o de a√ß√£o (CTA)

```html
<div class="text-center py-12">
    <p class="text-gray-400 mb-4">Nenhuma aposta realizada</p>
    <button class="bg-blue-600 text-white px-6 py-2 rounded">
        Fazer Aposta
    </button>
</div>
```

#### Error States
- [ ] Cor vermelha para erros (`text-red-400`, `bg-red-900`)
- [ ] Mensagem clara e acion√°vel

---

### 8. **Acessibilidade (WCAG)**
- [ ] Contraste m√≠nimo 4.5:1 (texto normal)
- [ ] Contraste m√≠nimo 3:1 (textos grandes)
- [ ] Labels em inputs (`<label>` ou `aria-label`)
- [ ] Navega√ß√£o por teclado funcional

---

### 9. **Performance Visual**
- [ ] Imagens otimizadas (WebP quando poss√≠vel)
- [ ] Lazy loading em listas longas
- [ ] Anima√ß√µes CSS (n√£o JS quando poss√≠vel)
- [ ] Transi√ß√µes suaves (`transition-all duration-200`)

---

### 10. **Consist√™ncia de Layout**
- [ ] Espa√ßamento uniforme (m√∫ltiplos de 4px/1rem)
- [ ] Grid/Flexbox para layouts
- [ ] Alinhamento consistente
- [ ] Hierarquia visual clara (t√≠tulos > subt√≠tulos > corpo)

---

## üö® Red Flags Cr√≠ticos

| Problema | Severidade | A√ß√£o |
|----------|-----------|------|
| Fundo claro (light mode) | üî¥ CR√çTICO | Mudar para dark |
| Cor hardcoded (#22c55e) | üî¥ CR√çTICO | Usar vari√°vel CSS |
| Fonte errada em t√≠tulos | üü† ALTO | Aplicar Russo One |
| Sem fallback de escudo | üü† ALTO | Adicionar onerror |
| Sem responsividade | üü° M√âDIO | Adicionar breakpoints |
| Contraste baixo | üü° M√âDIO | Ajustar cores |

---

## üìä Exemplo Completo (Header de M√≥dulo)

```html
<!-- Header Artilheiro Campe√£o -->
<div class="bg-gray-800 rounded-lg shadow-lg overflow-hidden">
    <!-- Header com gradiente do m√≥dulo -->
    <div class="p-6"
         style="background: var(--gradient-artilheiro);
                border-bottom: 2px solid var(--module-artilheiro-border);">
        <h1 class="font-russo text-3xl text-white mb-2">
            Artilheiro Campe√£o
        </h1>
        <p class="font-inter text-gray-200">
            Apostas no artilheiro da temporada
        </p>
    </div>

    <!-- Conte√∫do -->
    <div class="p-6">
        <div class="grid grid-cols-1 md:grid-cols-3 gap-4">
            <!-- Card estat√≠stica -->
            <div class="bg-gray-700 rounded-lg p-4">
                <p class="font-inter text-gray-400 text-sm mb-1">
                    Total de Apostas
                </p>
                <p class="font-jetbrains text-3xl text-white">
                    127
                </p>
            </div>
        </div>
    </div>
</div>
```

---

## üîó Refer√™ncias
- `CLAUDE.md` ‚Üí Se√ß√£o "UI/UX Guidelines"
- `/css/_admin-tokens.css` ‚Üí Vari√°veis CSS dos m√≥dulos
- `public/js/admin/*-management.js` ‚Üí Implementa√ß√µes de refer√™ncia

---

## 11. **Sistema de Notifica√ß√µes e Avisos**

### Estrutura Dual (Admin + Participante)

**Admin - M√≥dulo Notificador (Opera√ß√µes)**
- [ ] Interface para criar/gerenciar avisos globais
- [ ] Segmenta√ß√£o por liga/participante/temporada
- [ ] Preview antes de publicar
- [ ] Hist√≥rico de avisos enviados
- [ ] Categorias: `info`, `warning`, `success`, `urgent`

**Participante - Se√ß√£o "Avisos" (Home)**
- [ ] Cards horizontais em scroll (hide-scrollbar)
- [ ] Badge de categoria com cores distintas
- [ ] Limite de 5 avisos na home
- [ ] Link "Ver todos" para tela dedicada
- [ ] Auto-dismissal ap√≥s 7 dias (configur√°vel)

### Estrutura de Card de Aviso

```html
<!-- Card Padr√£o -->
<div class="aviso-card bg-gray-800 rounded-xl min-w-[240px] p-4 border-l-4"
     data-categoria="success">
    <div class="flex items-center gap-2 mb-2">
        <span class="material-icons text-green-500">check_circle</span>
        <p class="font-russo text-sm uppercase text-white">Rodada Confirmada</p>
    </div>
    <p class="text-gray-400 text-sm">
        Sua escala√ß√£o para o cl√°ssico foi salva com sucesso!
    </p>
</div>
```

### Cores por Categoria

| Categoria | Border Color | Icon Color | Uso |
|-----------|--------------|------------|-----|
| **success** | `border-green-500` | `text-green-500` | Confirma√ß√µes, sucesso |
| **warning** | `border-yellow-500` | `text-yellow-500` | Alertas, prazos |
| **info** | `border-blue-500` | `text-blue-500` | Informa√ß√µes gerais |
| **urgent** | `border-red-500` | `text-red-500` | A√ß√µes cr√≠ticas, erros |

### Checklist de Implementa√ß√£o

- [ ] Criar collection `avisos` no MongoDB
- [ ] Schema: `{ titulo, mensagem, categoria, ligaId?, timeId?, dataExpiracao, lido: Boolean }`
- [ ] Endpoint admin: `POST /api/admin/avisos/criar`
- [ ] Endpoint participante: `GET /api/avisos?ligaId={id}&timeId={id}`
- [ ] Marcar como lido: `POST /api/avisos/marcar-lido/:avisoId`
- [ ] UI admin em `/admin/operacoes/notificador`
- [ ] Componente `<AvisosList>` na home do participante

---

## 12. **Otimiza√ß√£o de Cards e Espa√ßamento**

### Princ√≠pio: Densidade Balanceada

**Objetivo:** Maximizar informa√ß√£o vis√≠vel sem sacrificar legibilidade (inspirado em dashboard mobile-first).

### Cards Compactos - Stats e M√©tricas

**Antes (Espa√ßoso):**
```html
<div class="bg-gray-800 rounded-lg p-6 shadow-lg">
    <h3 class="text-xl mb-4">Saldo Financeiro</h3>
    <p class="text-3xl">R$ 150,00</p>
</div>
```

**Depois (Compacto):**
```html
<div class="bg-gray-800 rounded-xl p-3 shadow-sm">
    <div class="flex items-center gap-2 mb-1">
        <span class="material-icons text-sm text-green-500">account_balance_wallet</span>
        <p class="text-[10px] font-bold uppercase text-gray-400 tracking-wider">Saldo</p>
    </div>
    <p class="font-jetbrains text-xl font-bold text-white">R$ 150</p>
</div>
```

**M√©tricas aplicadas:**
- Padding reduzido: `p-6` ‚Üí `p-3`
- T√≠tulo menor: `text-xl` ‚Üí `text-[10px] uppercase`
- √çcone integrado (14-16px)
- Border-radius aumentado: `rounded-lg` ‚Üí `rounded-xl` (visual moderno)
- Shadow sutil: `shadow-lg` ‚Üí `shadow-sm`

### Grid de Stats - Mobile First

**4 Cards compactos (2x2 em mobile):**
```html
<div class="grid grid-cols-2 gap-3">
    <!-- Card 1: Saldo -->
    <div class="bg-gray-800 rounded-xl p-3">
        <div class="flex items-center gap-1 mb-1">
            <span class="material-icons text-green-500 text-sm">account_balance_wallet</span>
        </div>
        <p class="font-jetbrains text-lg font-bold">R$ 150</p>
        <p class="text-[9px] uppercase text-gray-500 tracking-wide">Saldo</p>
    </div>

    <!-- Card 2: Posi√ß√£o -->
    <div class="bg-gray-800 rounded-xl p-3">
        <div class="flex items-center gap-1 mb-1">
            <span class="material-icons text-orange-500 text-sm">emoji_events</span>
        </div>
        <p class="font-jetbrains text-lg font-bold">#3</p>
        <p class="text-[9px] uppercase text-gray-500 tracking-wide">Posi√ß√£o</p>
    </div>

    <!-- Card 3: Pontos Rodada -->
    <!-- Card 4: Falta -->
</div>
```

### Checklist de Otimiza√ß√£o

- [ ] **Padding:** Reduzir de `p-6` para `p-3` ou `p-4`
- [ ] **Gap:** Grid com `gap-3` (12px) ao inv√©s de `gap-6`
- [ ] **Tipografia Labels:** `text-[10px] uppercase tracking-wider`
- [ ] **N√∫meros:** Manter leg√≠veis (`text-xl` ou `text-2xl`)
- [ ] **√çcones:** 14-16px (class `text-sm`)
- [ ] **Altura m√≠nima:** Remover `min-h-[150px]` de cards stat
- [ ] **Grid responsivo:** `grid-cols-2 md:grid-cols-4`

### Hierarquia Visual (Escala)

| Elemento | Tamanho | Uso |
|----------|---------|-----|
| **Hero Stats** | `text-4xl` | Pontua√ß√£o principal, destaque |
| **Card Stats** | `text-xl` / `text-2xl` | M√©tricas importantes |
| **Labels** | `text-[10px]` uppercase | T√≠tulos de cards |
| **Descri√ß√µes** | `text-xs` / `text-sm` | Subtextos |
| **Captions** | `text-[9px]` | Metadados, timestamps |

---

## 13. **Padr√£o de Carregamento Instant√¢neo**

**Inspirado em:** Dashboard Sa√∫de (IndexedDB cache-first)

### Estrat√©gia Cache-First

```javascript
// 1. Renderizar cache imediatamente (0ms)
const cached = await loadFromCache();
if (cached) renderUI(cached);

// 2. Buscar dados frescos em background
const fresh = await fetch('/api/data');
saveToCache(fresh);
renderUI(fresh); // Atualiza silenciosamente
```

### Checklist

- [ ] Implementar IndexedDB para dados cr√≠ticos (ranking, rodadas, extrato)
- [ ] TTL configur√°vel por tipo de dado (30s-5min)
- [ ] Skeleton loading APENAS na primeira visita (sem cache)
- [ ] Indicador visual sutil quando atualiza em background
- [ ] Invalida√ß√£o inteligente (rodada finalizada = limpar cache)

---

**√öltima atualiza√ß√£o:** 04/02/2026
**Vers√£o:** 1.1.0
